WordPress/wordpress-playground
# Support CORS proxy rate-limiting

## Motivation for the change, related issues

We want to prevent abuse of the CORS proxy, so we are:
- Adding a way to implement rate-limiting
- Requiring admins to explicitly declare whether to rate-limit or disable rate-limiting
- Implementing an initial rate-limiting scheme to be used on playground.wordpress.net.

We could also consider response caching. This might even be something we get for free for GET requests via the WP Cloud edge cache.

## Implementation details

If a `playground_cors_proxy_maybe_rate_limit()` function exists, it is called before proceeding to make a proxied HTTP request. The function is completely responsible for what it does, how it rate-limits, and what response is sent to the client.

Our implementation for rate-limiting on WP Cloud uses a [token bucket](https://en.wikipedia.org/wiki/Token_bucket) algorithm and tracks tokens for specific IPs in MySQL.

## Testing Instructions (or ideally a Blueprint)

Manual testing:
- Create a local DB with the cors_proxy_rate_limiting table.
- Temporarily define DB-related constants to use the database.
- Temporarily set default rate-limiting bucket to small values like `capacity: 5, fill_rate_per_minute: 5`
- Copy the rate-limiting PHP files to a test directory including `cors-proxy-config.php`
- cd to the directory
- Run a PHP server with `php -S localhost:7779 -t .`
- Use a browser to visit https://localhost:7779/cors-proxy.php?https://wordpress.org, refresh 5 times, and confirm the sixth request is rate-limited.
- Wait 12 seconds so a single request token is filled, refresh again, and confirm the request is not rate-limited.
- Refresh again and confirm the request is rate-limited again (because there are again no tokens remaining)
